src/value.rs

```diff
@@ -186,3 +189,267 @@ impl From<Bytes> for UIntValue {
         Self::try_from(value.0.as_ref()).expect("At most 32 bytes")
     }
 }
+
+/// Various value constructors.
+pub trait ValueConstructible: Sized + From<Option<Self>> + From<bool> + From<UIntValue> {
+    /// Create the unit value.
+    fn unit() -> Self;
+
+    /// Create the left value `Either::Left(inner)`.
+    fn left(inner: Self) -> Self;
+
+    /// Create the right value `Either::Right(inner)`.
+    fn right(inner: Self) -> Self;
+
+    /// Create the product value `(left, right)`.
+    fn product(left: Self, right: Self) -> Self;
+
+    /// Create an array from the given `elements`.
+    ///
+    /// ## Nonemptiness
+    ///
+    /// There must be at least one element.
+    ///
+    /// ## Errors
+    ///
+    /// Returns [`None`] if there are no elements.
+    // TODO: Enforce the nonemptiness of the iterator through the type system?
```

Comment:
In 892062c29dcc3787f20539b31a7bc153c29ae507
 
I'd just drop this TODO (and the one below). I've tried to enforce conditions like this with the type system and it just winds up being super unergonomic in Rust.

Reply:
In theory any property of a value can be a type, but Rust's type system has its limits.